src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/Processor.java

```diff
@@ -60,31 +69,76 @@ protected void processParsedEnvelopeDocuments(
         deleteBlob(envelope, cloudBlockBlob);
     }
 
+    protected void handleEventRelatedError(Event event, String containerName, String zipFilename, Exception exception) {
+        registerEvent(event, containerName, zipFilename, exception.getMessage());
+        log.error(exception.getMessage(), exception);
+    }
+
+    protected void registerEvent(Event event, String container, String zipFileName, String reason) {
+        ProcessEvent processEvent = new ProcessEvent(
+            container,
+            zipFileName,
+            event
+        );
+
+        processEvent.setReason(reason);
+        eventRepository.save(processEvent);
+
+        log.info(
+            "Zip {} from {} marked as {}",
+            processEvent.getZipFileName(),
+            processEvent.getContainer(),
+            processEvent.getEvent()
+        );
+    }
+
     private Boolean uploadParsedEnvelopeDocuments(
         Envelope envelope,
         List<Pdf> pdfs
     ) {
-        return errorWrapper.wrapDocUploadFailure(envelope, () -> {
+        try {
             documentProcessor.uploadPdfFiles(pdfs, envelope.getScannableItems());
             return Boolean.TRUE;
-        });
+        } catch (Exception ex) {
+            incrementUploadFailureCount(envelope);
+            updateEnvelopeLastStatus(envelope, Event.DOC_UPLOAD_FAILURE);
+            handleEventRelatedError(Event.DOC_UPLOAD_FAILURE, envelope.getContainer(), envelope.getZipFileName(), ex);
+            return Boolean.FALSE;
+        }
+    }
+
+    private void incrementUploadFailureCount(Envelope envelope) {
+        envelope.setUploadFailureCount(envelope.getUploadFailureCount() + 1);
```

Comment:
hmmm maybe we should consider this should be in the envelope.
Id also consider putting the envelope repository reference into the envelope too.
This way you would say 

envelope.failed(status)
envelope.save()

The envelope failure could increment a metric that is injected into the envelope.
This would require an envelope factory too.

Reply:
Or even hide the save altogether ...